# HW Week12

In [258]:
import math
import numpy as np
import pandas as pd
import quandl
import datetime
from scipy.stats import norm

### 10.1 Let S=\\$100, K=\\$105, r=8%, T=0.5, and δ=0.  Let u=1.3, d=0.8, and n=1 <br>
a. What are the premium, Δ, and B for a European call


In [259]:
S = 100
K = 105
r = 0.08
T = 0.5
st_dev = 0
u = 1.3
d = 0.8
n = 1

In [260]:
p = (math.exp(r*T) - d)/(u-d)
p

0.48162154838477633

In [261]:
premium = (math.exp(-r*T)*(p*max(S*u-K,st_dev)) + ((1-p)*max(S*d-K,st_dev)))
premium

11.568422433907068

In [262]:
delta = math.exp(-r*st_dev)*(max(S*u - K,st_dev) - max(S*d - K,st_dev))/(S*(u-d))
delta

0.5

In [263]:
B = premium - delta*S
B

-38.43157756609293

b. What are the premium, Δ, and B for a European put

In [264]:
p = (math.exp(r*T) - d)/(u-d)
p

0.48162154838477633

In [265]:
premium = (math.exp(-r*T)*(p*max(K-S*u,st_dev)) + ((1-p)*max(K-S*d,st_dev)))
premium

12.959461290380592

In [266]:
delta = math.exp(-r*st_dev)*(max(K-S*u,st_dev) - max(K-S*d,st_dev))/(S*(u-d))
delta

-0.5

In [267]:
B = premium - delta*S
B

62.95946129038059

### 10.4 Obtain at least 5 years’ worth of daily or weekly stock price data for a stock of your choice.



In [268]:
#daily stock prices over the past five years, starting at April 20, 2014
start = datetime.datetime(2014,4,1)
end = datetime.date.today()
 
#Apple stock data
apple = quandl.get("WIKI/" + "AAPL", start_date=start, end_date=end)

apple.head()

Open    High     Low   Close      Volume  Ex-Dividend  \
Date                                                                  
2014-04-01  537.76  541.87  536.77  541.65   7170000.0          0.0   
2014-04-02  542.38  543.48  540.26  542.55   6443600.0          0.0   
2014-04-03  541.39  542.50  537.64  538.79   5798000.0          0.0   
2014-04-04  539.81  540.00  530.58  531.82   9830400.0          0.0   
2014-04-07  528.02  530.90  521.89  523.47  10351800.0          0.0   

            Split Ratio  Adj. Open  Adj. High   Adj. Low  Adj. Close  \
Date                                                                   
2014-04-01          1.0  71.905792  72.455354  71.773415   72.425937   
2014-04-02          1.0  72.523548  72.670633  72.240076   72.546280   
2014-04-03          1.0  72.391172  72.539594  71.889746   72.043517   
2014-04-04          1.0  72.179904  72.205310  70.945729   71.111533   
2014-04-07          1.0  70.603422  70.988517  69.783758   69.995025   

            Adj. Volume  
Date                     
2014-04-01   50190000.0  
2014-04-02   45105200.0  
2014-04-03   40586000.0  
2014-04-04   68812800.0  
2014-04-07   72462600.0

1. Compute annual volatility using all the data.

In [269]:
# compute daily returns
daily_returns = apple['Adj. Close'].pct_change().dropna().values.T 
# compute daily volatility  
historical_vol_daily = np.std(daily_returns)  
# convert daily volatility to annual volatility, assuming 252 trading days  
historical_vol_annually = historical_vol_daily*np.sqrt(252)  
# return estimate of annual volatility  
print('Annual volatility equals to ' + str(100*historical_vol_annually) + ' %')

Annual volatility equals to 22.81207108836306 %


2. Compute annual volatility for each calendar year in your data. How does
volatility vary over time?

In [270]:
#2014
close = apple['2014-01':'2014-12']['Adj. Close']
r = np.diff(np.log(close))
r_mean = np.mean(r)
diff_square = [(r[i]-r_mean)**2 for i in range(0,len(r))]
std = np.sqrt(sum(diff_square)*(1.0/(len(r)-1)))
vol_2014 = std*np.sqrt(252)


#2015
close = apple['2015-01':'2015-12']['Adj. Close']
r = np.diff(np.log(close))
r_mean = np.mean(r)
diff_square = [(r[i]-r_mean)**2 for i in range(0,len(r))]
std = np.sqrt(sum(diff_square)*(1.0/(len(r)-1)))
vol_2015 = std*np.sqrt(252)


#2016
close = apple['2016-01':'2016-12']['Adj. Close']
r = np.diff(np.log(close))
r_mean = np.mean(r)
diff_square = [(r[i]-r_mean)**2 for i in range(0,len(r))]
std = np.sqrt(sum(diff_square)*(1.0/(len(r)-1)))
vol_2016 = std*np.sqrt(252)


#2017
close = apple['2017-01':'2017-12']['Adj. Close']
r = np.diff(np.log(close))
r_mean = np.mean(r)
diff_square = [(r[i]-r_mean)**2 for i in range(0,len(r))]
std = np.sqrt(sum(diff_square)*(1.0/(len(r)-1)))
vol_2017 = std*np.sqrt(252)


#2018
close = apple['2018-01':'2018-12']['Adj. Close']
r = np.diff(np.log(close))
r_mean = np.mean(r)
diff_square = [(r[i]-r_mean)**2 for i in range(0,len(r))]
std = np.sqrt(sum(diff_square)*(1.0/(len(r)-1)))
vol_2018 = std*np.sqrt(252)

pd.DataFrame({'year':[2014,2015,2016,2017,2018],'volatility':[vol_2014,vol_2015,vol_2016,vol_2017,vol_2018]})

year  volatility
0  2014    0.209172
1  2015    0.267676
2  2016    0.234186
3  2017    0.177260
4  2018    0.266811

3. Compute annual volatility for the first and second half of each year in your
data. How much variation is there in your estimate?

In [271]:
#2014 first half
close = apple['2014-01':'2014-6']['Adj. Close']
r = np.diff(np.log(close))
r_mean = np.mean(r)
diff_square = [(r[i]-r_mean)**2 for i in range(0,len(r))]
std = np.sqrt(sum(diff_square)*(1.0/(len(r)-1)))
vol_2014_1 = std*np.sqrt(252/2)

#2014 second half
close = apple['2014-07':'2014-12']['Adj. Close']
r = np.diff(np.log(close))
r_mean = np.mean(r)
diff_square = [(r[i]-r_mean)**2 for i in range(0,len(r))]
std = np.sqrt(sum(diff_square)*(1.0/(len(r)-1)))
vol_2014_2 = std*np.sqrt(252/2)

#2015 first half
close = apple['2015-01':'2015-6']['Adj. Close']
r = np.diff(np.log(close))
r_mean = np.mean(r)
diff_square = [(r[i]-r_mean)**2 for i in range(0,len(r))]
std = np.sqrt(sum(diff_square)*(1.0/(len(r)-1)))
vol_2015_1 = std*np.sqrt(252/2)

#2015 second half
close = apple['2015-07':'2015-12']['Adj. Close']
r = np.diff(np.log(close))
r_mean = np.mean(r)
diff_square = [(r[i]-r_mean)**2 for i in range(0,len(r))]
std = np.sqrt(sum(diff_square)*(1.0/(len(r)-1)))
vol_2015_2 = std*np.sqrt(252/2)

#2016 first half
close = apple['2016-01':'2016-6']['Adj. Close']
r = np.diff(np.log(close))
r_mean = np.mean(r)
diff_square = [(r[i]-r_mean)**2 for i in range(0,len(r))]
std = np.sqrt(sum(diff_square)*(1.0/(len(r)-1)))
vol_2016_1 = std*np.sqrt(252/2)

#2016 second half
close = apple['2016-07':'2016-12']['Adj. Close']
r = np.diff(np.log(close))
r_mean = np.mean(r)
diff_square = [(r[i]-r_mean)**2 for i in range(0,len(r))]
std = np.sqrt(sum(diff_square)*(1.0/(len(r)-1)))
vol_2016_2 = std*np.sqrt(252/2)

#2017 first half
close = apple['2017-01':'2017-6']['Adj. Close']
r = np.diff(np.log(close))
r_mean = np.mean(r)
diff_square = [(r[i]-r_mean)**2 for i in range(0,len(r))]
std = np.sqrt(sum(diff_square)*(1.0/(len(r)-1)))
vol_2017_1 = std*np.sqrt(252/2)

#2017 second half
close = apple['2017-07':'2017-12']['Adj. Close']
r = np.diff(np.log(close))
r_mean = np.mean(r)
diff_square = [(r[i]-r_mean)**2 for i in range(0,len(r))]
std = np.sqrt(sum(diff_square)*(1.0/(len(r)-1)))
vol_2017_2 = std*np.sqrt(252/2)

#2018 first half
close = apple['2018-01':'2018-6']['Adj. Close']
r = np.diff(np.log(close))
r_mean = np.mean(r)
diff_square = [(r[i]-r_mean)**2 for i in range(0,len(r))]
std = np.sqrt(sum(diff_square)*(1.0/(len(r)-1)))
vol_2018_1 = std*np.sqrt(252/2)

#2014 second half
close = apple['2018-07':'2018-12']['Adj. Close']
r = np.diff(np.log(close))
r_mean = np.mean(r)
diff_square = [(r[i]-r_mean)**2 for i in range(0,len(r))]
std = np.sqrt(sum(diff_square)*(1.0/(len(r)-1)))
vol_2018_2 = std*np.sqrt(252/2)

pd.DataFrame({'year':["2014 fist half","2014 second half","2015 fist half","2015 second half",
                      "2016 fist half","2016 second half","2017 fist half","2017 second half",
                      "2018 fist half","2018 second half"],
              'volatility':[vol_2014_1,vol_2014_2,vol_2015_1,vol_2015_2,
                            vol_2016_1,vol_2016_2, vol_2017_1,vol_2017_2,
                            vol_2018_1,vol_2018_2]})

/Users/FOMIOLNY/anaconda3/lib/python3.6/site-packages/numpy/core/fromnumeric.py:2920: RuntimeWarning: Mean of empty slice.
  out=out, **kwargs)
/Users/FOMIOLNY/anaconda3/lib/python3.6/site-packages/numpy/core/_methods.py:85: RuntimeWarning: invalid value encountered in double_scalars
  ret = ret.dtype.type(ret / rcount)


year  volatility
0    2014 fist half    0.157105
1  2014 second half    0.143981
2    2015 fist half    0.164184
3  2015 second half    0.210905
4    2016 fist half    0.192059
5  2016 second half    0.134962
6    2017 fist half    0.121083
7  2017 second half    0.130333
8    2018 fist half    0.188664
9  2018 second half   -0.000000

### 11.1 Consider a one-period binomial model with h = 1, where S = \\$100, r = 0, σ = 0.3, δ = 0.08. Compute American call option prices for K = \\$70, \\$80, \\$90, and \\$100.



In [272]:
S = 100
r = 0
T = 0.5
h = 1
st_dev = 0.3
delta = 0.08
r = 0

a. At which strike(s) does early exercise occur?

In [273]:
u = math.exp((r-delta)+(st_dev*np.sqrt(h)))
d = math.exp((r-delta)-(st_dev*np.sqrt(h)))
p = (math.exp(r-delta)-d)/(u-d)
def strike_call(k):
    if uS-K>0:
       part1= p*(uS-K)
    else: part1=0
    if dS-K>0:
       part2= p*(dS-K)
    else: part2=0 
    strike_call = round(math.exp(-r*h)*(part1+part2),2)
    print("The value of the one period binomial European $"+str(k)+" strike call is " + str(strike_call))
    if 100-k > strike_call:
       print("Early exercise occurs since the value of immediate exercise $"+str(100-k)+" is greater than the strike call of " + str(strike_call))  
    else:
       print("Early exercise doesn't occur since the value of immediate exercise $"+str(100-k)+" is greater than the strike call of " + str(strike_call))  

In [274]:
K = 70
strike_call(K)

The value of the one period binomial European $70 strike call is 23.24
Early exercise occurs since the value of immediate exercise $30 is greater than the strike call of 23.24


In [275]:
K = 80
strike_call(K)

The value of the one period binomial European $80 strike call is 18.98
Early exercise occurs since the value of immediate exercise $20 is greater than the strike call of 18.98


In [276]:
K = 90
strike_call(K)

The value of the one period binomial European $90 strike call is 14.73
Early exercise doesn't occur since the value of immediate exercise $10 is greater than the strike call of 14.73


In [277]:
K = 100
strike_call(K)

The value of the one period binomial European $100 strike call is 10.47
Early exercise doesn't occur since the value of immediate exercise $0 is greater than the strike call of 10.47


b. Use put-call parity to explain why early exercise does not occur at the higher
strikes.

In [278]:
S = 100
#accorting to put-call parity C = S*math.exp(-delta)-K*math.exp(-r)+P
print(S*math.exp(-delta))
print(math.exp(-r))

92.31163463866358
1.0


From the equation above we get 92.31-K+P. If 100-K is larger than 92.31-K+P the option can be exercised early. As at the higher strikes like \\$90 immideate exercise price  100-90=10 is smaller than 92.31-90+P = 2.31-P early exercise doesn't occcur.

### 12.3 Let S = \\$100, K = $120, σ = 30%, r = 0.08, and δ = 0.
a. Compute the Black-Scholes call price for 1 year to maturity and for a variety
of very long times to maturity. What happens to the option price as T → ∞?

In [279]:
S = 100
K = 120
r = 0.08
delta = 0
st_dev = 0.3

In [280]:
#at T=1
T=1
d1_1 = (math.log(S/K) + (r - delta + 0.5*pow(st_dev,2))*T)/(st_dev*np.sqrt(T))
d2_1 = d1_1-st_dev*math.sqrt(T)
BS_call_price_1 = S*math.exp(-delta*T)*norm.cdf(d1_1) - K*math.exp(-r*T)*norm.cdf(d2_1)

#at T=2
T=2
d1_2 = (math.log(S/K) + (r - delta + 0.5*pow(st_dev,2))*T)/(st_dev*np.sqrt(T))
d2_2 = d1_1-st_dev*math.sqrt(T)
BS_call_price_2 = S*math.exp(-delta*T)*norm.cdf(d1_2) - K*math.exp(-r*T)*norm.cdf(d2_2)

#at T=5
T=5
d1_5 = (math.log(S/K) + (r - delta + 0.5*pow(st_dev,2))*T)/(st_dev*np.sqrt(T))
d2_5 = d1_1-st_dev*math.sqrt(T)
BS_call_price_5 = S*math.exp(-delta*T)*norm.cdf(d1_5) - K*math.exp(-r*T)*norm.cdf(d2_5)

#at T=10
T=10
d1_10 = (math.log(S/K) + (r - delta + 0.5*pow(st_dev,2))*T)/(st_dev*np.sqrt(T))
d2_10 = d1_1-st_dev*math.sqrt(T)
BS_call_price_10 = S*math.exp(-delta*T)*norm.cdf(d1_10) - K*math.exp(-r*T)*norm.cdf(d2_10)

#at T=50
T=50
d1_50 =(math.log(S/K) + (r - delta + 0.5*pow(st_dev,2))*T)/(st_dev*np.sqrt(T))
d2_50 = d1_1-st_dev*math.sqrt(T)
BS_call_price_50 = S*math.exp(-delta*T)*norm.cdf(d1_50) - K*math.exp(-r*T)*norm.cdf(d2_50)

#at T=100
T=100
d1_100 = (math.log(S/K) + (r - delta + 0.5*pow(st_dev,2))*T)/(st_dev*np.sqrt(T))
d2_100 = d1_1-st_dev*math.sqrt(T)
BS_call_price_100 = S*math.exp(-delta*T)*norm.cdf(d1_100) - K*math.exp(-r*T)*norm.cdf(d2_100)

#at T=500
T=500
d1_500 =(math.log(S/K) + (r - delta + 0.5*pow(st_dev,2))*T)/(st_dev*np.sqrt(T))
d2_500 = d1_1-st_dev*math.sqrt(T)
BS_call_price_500 = S*math.exp(-delta*T)*norm.cdf(d1_500) - K*math.exp(-r*T)*norm.cdf(d2_500)

pd.DataFrame({'T':["d1","d2","BS Call Price"],
              "1":[d1_1,d2_1,BS_call_price_1],
              "2":[d1_2,d2_2,BS_call_price_2],
              "5":[d1_5,d2_5,BS_call_price_5],
              "10":[d1_10,d2_10,BS_call_price_10],
              "50":[d1_50,d2_50,BS_call_price_50],
              "100":[d1_100,d2_100,BS_call_price_100],
              "500":[d1_500,d2_500,BS_call_price_500]})

T         1          2          5         10         50  \
0             d1 -0.191072   0.159520   0.659906   1.125432   2.860331   
1             d2 -0.491072  -0.615336  -0.861892  -1.139755  -2.312392   
2  BS Call Price  7.896571  28.812803  58.899207  80.121454  99.765591   

         100         500  
0   4.105893    9.289771  
1  -3.191072   -6.899276  
2  99.997958  100.000000

As T approaches infinity, the call approaches the value of the underlying stock price, signifying that over very long time horizons the call option is not distinguishable from the stock.

### 13.3 In the following problems assume, unless otherwise stated, that S = \\$40, σ = 30%, r = 8%, and δ = 0. Suppose you buy a 40–45 bull spread with 91 days to expiration. If you delta-hedge this position, what investment is required? What is your overnight profit if the stock tomorrow is \\$39? What if the stock is \\$40.50?

In [281]:
S = 40
K = 45
r = 0.08
delta = 0
st_dev = 0.3

In [282]:
#91st day
T = 91/365
d1 = (math.log(S/K) + (r - delta + 0.5*pow(st_dev,2))*T)/(st_dev*np.sqrt(T))
print("d1 is "+str(d1))
d2 = d1-st_dev*math.sqrt(T)
print("d2 is "+str(d2))
hedge_delta = norm.cdf(d1)
print("delta is "+str(hedge_delta))
#as we are trying to hedge a sold option we replicate the long option by buying delta shares. 
BS_call_price = S*math.exp(-delta*T)*norm.cdf(d1) - K*math.exp(-r*T)*norm.cdf(d2)
print("Selling call, getting premium "+str(BS_call_price))
print("Buying delta shares "+ str(S*hedge_delta))
b_netcost = abs(S*hedge_delta) - BS_call_price
print("Borrowing the net cost " + str(b_netcost))
#no net overflow per share

d1 is -0.5782503476905758
d2 is -0.728044727306046
delta is 0.2815475557049747
Selling call, getting premium 0.971026784192718
Buying delta shares 11.261902228198988
Borrowing the net cost 10.29087544400627


In [283]:
#90th day, if price is $39
S = 39
K = 45
T = 90/365
d1 = (math.log(S/K) + (r - delta + 0.5*pow(st_dev,2))*T)/(st_dev*np.sqrt(T))
print("d1 is "+str(d1))
d2 = d1-st_dev*math.sqrt(T)
print("d2 is "+str(d2))
#as we are trying to hedge a sold option we replicate the long option by buying delta shares. 
BS_call_price = S*math.exp(-delta*T)*norm.cdf(d1) - K*math.exp(-r*T)*norm.cdf(d2)
print("Liability under option "+ str(BS_call_price))
print("Value of shares held "+ str(abs(hedge_delta)*S))
under_loan = -b_netcost*math.exp(r*(1/365))
print("Money owed under loan "+ str(under_loan))

d1 is -0.753706345944088
d2 is -0.9026754058925143
Liability under option 0.7054482611978461
Value of shares held 10.980354672494013
Money owed under loan -10.293131225550068


In [284]:
#90th day, if price is $40.50
S = 40.50
K = 45
T = 90/365
d1 = (math.log(S/K) + (r - delta + 0.5*pow(st_dev,2))*T)/(st_dev*np.sqrt(T))
print("d1 is "+str(d1))
d2 = d1-st_dev*math.sqrt(T)
print("d2 is "+str(d2))
#as we are trying to hedge a sold option we replicate the long option by buying delta shares. 
BS_call_price = S*math.exp(-delta*T)*norm.cdf(d1) - K*math.exp(-r*T)*norm.cdf(d2)
print("Liability under option "+ str(BS_call_price))
print("Value of shares held "+ str(abs(hedge_delta)*S))
under_loan = -b_netcost*math.exp(r*(1/365))
print("Money owed under loan "+ str(under_loan))

d1 is -0.5003629470133779
d2 is -0.6493320069618042
Liability under option 1.104639661307809
Value of shares held 11.402676006051475
Money owed under loan -10.293131225550068


### 18.2 You draw these five numbers from a standard normal distribution:{−1.7, 0.55, −0.3,−0.02, .85}. What are the equivalent draws from a normal distribution with mean 0.8 and variance 25?

In [285]:
mean = 0.8
var = 25

I used the equation below to convert “Standard” normally distributed values to normally distributed values<br>
Y =(X - mean (X)) /SD(X), <br>
where X is normally distributed and Y is “standard” normally distributed<br>
X = Y*SD(X) + mean (X)

In [286]:
#convert −1.7
st_dev = math.sqrt(var)
X = -1.7*st_dev  + mean
X

-7.7

In [287]:
#convert 0.55
X = 0.55*st_dev  + mean
X

3.55

In [288]:
#convert −0.3
X = -0.3*st_dev  + mean
X

-0.7

In [289]:
#convert −0.02
X = -0.02*st_dev  + mean
X

0.7000000000000001

In [290]:
#convert 0.85
X = 0.85*st_dev  + mean
X

5.05

In [291]:
math.exp(-0.08)*70

64.6181442470645

In [292]:
math.exp(0.08)*70

75.83009473724711